File: apps/processor/src/events.ts
import type { NatsConnection } from "nats";

import type { DocumentEvent, DocumentLifecycleEvent } from "./types.js";

export const DOCUMENT_EVENTS_STREAM_NAME = "document-events";
export const DOCUMENT_EVENTS_STREAM_SUBJECT = "documents.events.>";
export const DOCUMENT_EVENT_SUBJECTS = {
	minioStored: "documents.events.minio.stored",
	processorQueued: "documents.events.processor.queued",
	processorStarted: "documents.events.processor.started",
	processorCompleted: "documents.events.processor.completed",
	processorFailed: "documents.events.processor.failed",
} as const;

export function buildLifecycleEvent(
	subject: string,
	event: DocumentEvent,
	overrides: Partial<DocumentLifecycleEvent> = {},
): DocumentLifecycleEvent {
	return {
		subject,
		documentId: event.documentId,
		bucket: event.bucket,
		objectKey: event.objectKey,
		contentType: event.contentType,
		processingVersion: event.processingVersion ?? 1,
		occurredAt: new Date().toISOString(),
		...overrides,
	};
}

export async function publishLifecycleEvent(
	nc: NatsConnection,
	event: DocumentLifecycleEvent,
): Promise<void> {
	nc.publish(event.subject, JSON.stringify(event));
	await nc.flush();
}

File: apps/pkg/documentevents/events.go
package documentevents

import "time"

const (
	StreamID             = "document-events"
	DefaultStreamName    = "document-events"
	DefaultStreamSubject = "documents.events.>"

	SubjectMinIOStored        = "documents.events.minio.stored"
	SubjectProcessorQueued    = "documents.events.processor.queued"
	SubjectProcessorStarted   = "documents.events.processor.started"
	SubjectProcessorCompleted = "documents.events.processor.completed"
	SubjectProcessorFailed    = "documents.events.processor.failed"
)

type LifecycleEvent struct {
	Subject           string `json:"subject"`
	DocumentID        string `json:"documentId"`
	Bucket            string `json:"bucket,omitempty"`
	ObjectKey         string `json:"objectKey,omitempty"`
	ContentType       string `json:"contentType,omitempty"`
	ProcessingVersion int    `json:"processingVersion,omitempty"`
	OccurredAt        string `json:"occurredAt"`
	Error             string `json:"error,omitempty"`
}

func NewLifecycleEvent(subject string, documentID string, bucket string, objectKey string, contentType string, processingVersion int) LifecycleEvent {
	return LifecycleEvent{
		Subject:           subject,
		DocumentID:        documentID,
		Bucket:            bucket,
		ObjectKey:         objectKey,
		ContentType:       contentType,
		ProcessingVersion: processingVersion,
		OccurredAt:        time.Now().UTC().Format(time.RFC3339),
	}
}
